plugins/artifacts/artifacts.go

```diff
@@ -0,0 +1,110 @@
+//go:generate go-composite-schema --unexported --required artifacts config-schema.yml generated_configschema.go
+
+// Package artifacts is responsible for uploading artifacts after builds
+package artifacts
+
+import (
+	"fmt"
+	"mime"
+	"path/filepath"
+
+	"github.com/taskcluster/taskcluster-client-go/tcclient"
+	"github.com/taskcluster/taskcluster-worker/engines"
+	"github.com/taskcluster/taskcluster-worker/plugins"
+	"github.com/taskcluster/taskcluster-worker/plugins/extpoints"
+	"github.com/taskcluster/taskcluster-worker/runtime"
+	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
+)
+
+type pluginProvider struct {
+	extpoints.PluginProviderBase
+}
+
+func (pluginProvider) NewPlugin(extpoints.PluginOptions) (plugins.Plugin, error) {
+	return plugin{}, nil
+}
+
+func (pluginProvider) ConfigSchema() runtime.CompositeSchema {
+	return runtime.NewEmptyCompositeSchema()
+}
+
+type plugin struct {
+	plugins.PluginBase
+}
+
+func (plugin) PayloadSchema() (runtime.CompositeSchema, error) {
+	return configSchema, nil
+}
+
+func (plugin) NewTaskPlugin(options plugins.TaskPluginOptions) (plugins.TaskPlugin, error) {
+	return &taskPlugin{
+		TaskPluginBase: plugins.TaskPluginBase{},
+		payload:        *(options.Payload.(*config)),
+	}, nil
+}
+
+type taskPlugin struct {
+	plugins.TaskPluginBase
+	context *runtime.TaskContext
+	payload config
+}
+
+func (tp *taskPlugin) Prepare(context *runtime.TaskContext) error {
+	tp.context = context
+	return nil
+}
+
+func (tp *taskPlugin) Stopped(result engines.ResultSet) (bool, error) {
+	for _, artifact := range tp.payload {
+		switch artifact.Type {
+		case "directory":
+			err := result.ExtractFolder(artifact.Path, tp.createUploadHandler(artifact.Name, artifact.Expires))
```

Comment:
This isn't enough... you need to handle `ErrHandlerInterrupt` which indicates that an upload failed... but the actual error returned by the handler is eaten, so you have store that in a separate variable... or something like that...